### PR TITLE
Add proxy audit log and other assorted fixes

### PR DESCRIPTION
**Pull request by scbedd**
This PR

- [x] Fixes an identified issue with not properly utilizing the session locks -- causing a serious perf degradation when adding a bunch of sanitizers to a bunch of tests. (like 10+ minutes per massive core job bad)
- [x] Fixes a serious problem that was allowing a test to be `stopped` while the last entry in the recording is still writing it's body. That location in the code has _already_ obtained a lock on the session, so merely adding the same to `playback/stop` prevents the test from being stopped while the last entry is being streamed back. This was the source of the `Channel Response timeout` issues that `java- core` was seeing @alzimmermsft (at least, it has not reproed in the 3 core runs since I added this fix 🤞 )
- [x] Add a concurrent queue for log messages that can be dumped. Currently this endlessly grows, but pulling with /Audit/GetAuditLogs will clean up that memory.
- [x] Add audit entries for `Record` code paths
- [x] Add a few tests for the new functionality
- [ ] Get signoff from @mikeharder about the always-on per-session audit logging. We were kinda laughing about chewing up memory, but...it's not that bad :P No affect to `java - core` which beats the absolute hell out of the server.

@mikeharder I need to add audit items to the `record` pathway, but I was laser focused on getting `playback` concurrency issues resolved for the java folks. Please take a look, see how I'm retaining the various locks on resources (by reducing to a single access lock to each playback session). I _think_ I've nailed most of the edge cases. 

@alzimmermsft looking at the latest `java - core` failures running with this version of the test-proxy, I don't think the proxy is related to those. I'm not declaring total victory with this PR but the issues are seriously reduced.

This version is standing up to multiple storage runs: 
image

### CODE REVIEW

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Properties/launchSettings.json [hunk 1]
```diff
@@ -2,6 +2,7 @@
   "profiles": {
     "Azure.Sdk.Tools.TestProxy": {
       "commandName": "Project",
+      "commandLineArgs": "--storage-location=C:/repo/azure-sdk-for-python",
```
Comment (scbedd) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Properties/launchSettings.json:5]:
```suggestion
```

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Audit.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            DebugLogger.LogAdminRequestDetails(_logger, Request);
+
+            var pathToAssets = RecordingHandler.GetAssetsJsonLocation(StoreResolver.ParseAssetsJsonBody(options), _recordingHandler.ContextDirectory);
+
+            await _recordingHandler.Store.Push(pathToAssets);
+        }
+
+        [HttpPost]
+        [AllowEmptyBody]
+        public async Task Stop([FromBody()] IDictionary<string, string> variables = null)
```
Comment (scbedd) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Audit.cs:65]: todo: cleanup random unused arguments. remove useless functions. etc.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs [hunk 3]
```diff
@@ -418,35 +462,58 @@ public async Task StartPlaybackAsync(string sessionId, HttpResponse outgoingResp
...
                 {
-                    throw new HttpException(HttpStatusCode.InternalServerError, $"Unexpectedly failed to remove in-memory session {session.SourceRecordingId}.");
+                    if (!InMemorySessions.TryGetValue(session.SourceRecordingId, out var inMemorySession))
+                    {
+                        throw new HttpException(HttpStatusCode.InternalServerError, $"Unexpectedly failed to retrieve in-memory session {session.SourceRecordingId}.");
+                    }
+
+                    Interlocked.Add(ref Startup.RequestsRecorded, -1 * inMemorySession.Session.Entries.Count);
+
+                    if (!InMemorySessions.TryRemove(session.SourceRecordingId, out _))
```
Comment (alzimmermsft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:503]: Couldn't just using `TryRemove(session.SourceRecordingId, out var inMemorySessions)` be sufficient here? Unless there are other locations outside of this EntryLock block which remove from the InMemorySessions which this catches a race condition between retrieval and removal.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs [hunk 4]
```diff
@@ -418,35 +462,58 @@ public async Task StartPlaybackAsync(string sessionId, HttpResponse outgoingResp
...
+                    {
+                        throw new HttpException(HttpStatusCode.InternalServerError, $"Unexpectedly failed to remove in-memory session {session.SourceRecordingId}.");
+                    }
+
+                    GC.Collect();
```
Comment (alzimmermsft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:508]: Is it common to force GC collection manually?

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Audit.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+            var allAuditSessions = _recordingHandler.RetrieveOngoingAuditLogs();
+            allAuditSessions.AddRange(_recordingHandler.AuditSessions.Values);
+
+            StringBuilder stringBuilder = new StringBuilder();
+
+            foreach (var auditLogQueue in allAuditSessions) {
+                while (auditLogQueue.TryDequeue(out var logItem))
+                {
+                    stringBuilder.Append(logItem.ToCsvString() + Environment.NewLine);
```
[the remark this answers is not in the document]
Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Audit.cs:43]: Yeah. It [literally doesn't work. ](https://stackoverflow.com/a/11155844) this confused dev had the same experience I did. Extremely frustrating.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs (hunk 4 above)
Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:508]: Not really, but it does work more effectively than not having it (at least on performance profiler).

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs (hunk 3 above)
Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:503]: I think you're right. It's not like we have an early exit. This will blow up if either of them fails. May as well reduce to one check.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs [hunk 6]
```diff
@@ -81,6 +82,48 @@ public class RecordingHandler
         public readonly ConcurrentDictionary<string, ModifiableRecordSession> PlaybackSessions
             = new ConcurrentDictionary<string, ModifiableRecordSession>();
 
+        // as sessions are popped off,
```
Comment (benbp) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:85]: incomplete comment?

**Review by mikeharder — APPROVED** [review]

**Review by benbp — APPROVED** [review]

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs (hunk 6 above)
Comment (semick-dev) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:85]: **deep sigh**

Comment (scbedd) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:85]:
```suggestion
```
